debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/processor/TransactionCommitConsumer.java

```diff
@@ -66,38 +71,26 @@
     private final BlockingConsumer<LogMinerEvent> delegate;
     private final OracleConnectorConfig connectorConfig;
     private final OracleDatabaseSchema schema;
-    private final List<String> lobWriteData;
-
-    /**
-     * Describes the current LOB event buffering state, whether we're working on a series of
-     * {@code LOB_WRITE} events, {@code LOB_ERASE} events, or any other type of event that
-     * does not require special handling.
-     */
-    enum LobState {
-        WRITE,
-        ERASE,
-        OTHER
-    };
-
-    private LogMinerEvent lastEvent;
-    private SelectLobLocatorEvent lastSelectLobLocatorEvent;
-    private LobState lobState;
+
+    private final Map<String, DmlEvent> rows = new HashMap<>();
+
+    private String currentLobRowId;
+    private String currentLobColumnName;
+    private int currentLobColumnPosition = -1;
 
     public TransactionCommitConsumer(BlockingConsumer<LogMinerEvent> delegate, OracleConnectorConfig connectorConfig, OracleDatabaseSchema schema) {
         this.delegate = delegate;
-        this.lobState = LobState.OTHER;
-        this.lobWriteData = new ArrayList<>();
         this.connectorConfig = connectorConfig;
         this.schema = schema;
     }
 
     @Override
     public void close() throws InterruptedException {
-        if (lastEvent != null) {
-            if (!lobWriteData.isEmpty()) {
-                mergeLobWriteData(lastEvent);
-            }
-            dispatchChangeEvent(lastEvent);
+        // sort the remaining events on increasing SCN before dispatch
+        List<DmlEvent> events = new ArrayList<DmlEvent>(rows.values());
+        Collections.sort(events, (a, b) -> a.getScn().compareTo(b.getScn()));
```

Comment:
I don't think we should try to implement any behavior based on sorting by system change numbers.  Oracle is allowed to emit a certain number of events with the same SCN and therefore we ultimately should rely on the order based on how they've been read as the source of record for their order than doing this.  Is there then any concern with this sort?

Reply:
The purpose of this sort was to make the records come out in a somewhat-predictable way. The implementation will typically emit all records from the `rows` map at the invocation of `TransactionCommitConsumer#close()` (because we hold on to the last "event batch" for each row until the TX is closed). That means that in a TX with lots of affected rows, the order in which they are emitted is probably totally unrelated to the order in which the events were read from LogMiner (the iteration order over the HashMap's value set will determine emission order). I thought the SCNs would be a good proxy for the "original" order, but I can make that original order explicit by adding in a counter in the row state, and sorting on that counter instead.

Reply:
I had the same inclination as well when I did some rework back in 1.4 or so and quickly found that applying any sort of order whether at the query level or based on objects in a collection caused problems.  In general, it's best to let LogMiner give us the order and we adhere to that.